src/substrate/filters/enricher.ts

```diff
@@ -101,55 +110,103 @@ export async function Enrich(
-        const currentEra = (Number)(await api.query.staking.currentEra<Option<EraIndex>>());
-        const eraPointsIndividual = validatorEraPoints.individual;
-        let active: Array<ValidatorId>
-        let waiting: Array<ValidatorId>
+
+        const validators = await api.query.session.validators.at(hash);
+        // get the era of block
+        const rawCurrentEra = await api.query.staking.currentEra.at(hash);
+        const currentEra = rawCurrentEra.toRawType
+          ? rawCurrentEra.toRawType() === 'u32'
+            ? rawCurrentEra.toString() as unknown as EraIndex
+            : rawCurrentEra.unwrap()
+          : rawCurrentEra.unwrap();
+        
+        // get the nextElected Validators
+        const keys = api.query.staking.erasStakers
+          ? await api.query.staking.erasStakers.keysAt(hash,currentEra)
+          : await api.query.staking.stakers.keysAt(hash, currentEra) ; 
+
+        const nextElected = keys.length ? keys.map((key) => key.args[1] as AccountId): validators['nextElected'];
+        
+        // get current stashes
+        const stashes = await api.query.staking.validators.keysAt(hash);
+
+        // find the waiting validator
+        const nextElectedStr = nextElected.map((v) => v.toString())
+        // const stashesStr =stashes.map((v) => v.args[0].toString())
+        const stashesStr =stashes.filter((v) => v.args.length).map((v) => v.args[0].toString())
+        
+        const waiting = stashesStr.filter((v) => !nextElectedStr.includes(v));
+
+
+        // get validators current era reward points
+        const validatorEraPoints: EraRewardPoints = await currentPoints(api, currentEra, hash, validators) as EraRewardPoints;
+        const eraPointsIndividual = validatorEraPoints.individual.toJSON();
+
         const validatorInfo = {};
-        electedInfo.info.forEach(async ({ accountId, controllerId, validatorPrefs, rewardDestination }) => {
-          const commissionPer = (Number)(validatorPrefs.commission || new BN(0)) / 10_000_000;
-          const key = accountId.toString();
-          validatorInfo[key] = {}
+
+        for(let validator of validators){
+          const key = validator.toString();
+
+          // eraValidatorPrefs does not return comission prior to 3139800
+          const preference = api.query.staking.erasValidatorPrefs
+            ? await api.query.staking.erasValidatorPrefs.at(hash, currentEra,key)
+            : await api.query.staking.validators.at(hash, key);
+
+          const commissionPer =  (Number)(preference.commission || new BN(0)) / 10_000_000;
+
+          const rewardDestination = await api.query.staking.payee.at(hash,key);
+          const controllerId = await api.query.staking.bonded.at(hash, key);
+          let nextSessionKeysOpt;
+
+          try{
+            // to get keys of old blocks
+            nextSessionKeysOpt = await api.query.session.nextKeys(hash,key);
```

Comment:
I don't understand how/why this line works. The first argument of `nextKeys` in v31 appears to be a fixed key, according to https://github.com/paritytech/substrate/blob/00a400f82539e2f78e8ddbcd98aea512c87c5f3c/frame/session/src/lib.rs#L362

So why pass in the block hash?

Reply:
Thanks for pointing out. Updated the api call for version 31 but after discussio with @drewstone on today's syncup, we don't need session keys in new session event so will be removing this piece of code.